### ISSUE TITLE
关于评估方式

### ISSUE THREAD

**Issue by ShellingFord221** [thunlp/OpenNRE#190]
你好！想问下基于bag-level的评估方式（计算precision, recall）和基于sentence-level的评估方式是否是等价的？假如我以sentence-level的方式训练模型，分别以bag-level方式和sentence-level方式进行评估，结果（precision, recall, AUC, F1）是否会有差异？如果有差异哪种方式指标会更高些？
谢谢！

**Comment by ShellingFord221** [comment]
还是说以bag-level建模就要以bag-level测试，以sentence-level建模就要以sentence-level测试，但两种方式得到的结果（precision, recall, AUC, F1）是具有可比性的？
希望能分别回答一下这2个问题，非常感谢！！

**Comment by gaotianyu1350** [comment]
两种方法得到的结果是没有可比性的：

1. 单位不一样。sentence-level以句为单位，bag-level以包为单位
2. sentence-level和bag-level的数据集也不一样。通常来说在什么数据集上就用什么评测

**Comment by ShellingFord221** [comment]
你好，想问下假如一个sample的标签是5，但是我预测的是3（假定0是NA类），那么这个sample是FP还是FN呢？谢谢！

**Comment by gaotianyu1350** [comment]
对于5这一类是false negative，对于3这一类是false positive，因为FP，FN是针对二分类而言的

**Comment by ShellingFord221** [comment]
你好，想问下：
1. CNN/PCNN+MAX/ATT是不是在train和test的时候都是一个bag（不管里面有多少个句子）只给出一个概率分布的表示，用于计算loss或者AUC呢？
2. 在Manual Evaluation的时候选取的top100, 200, 500 extracted instances是不是指的也是bag呢？就是根据概率选取排在前100的bag，然后手工去看这个bag到底应该是什么标签
谢谢！

**Comment by ShellingFord221** [comment]
你好，CNN/PCNN+MAX/ATT这种都是基于bag来train和test的，如果bag有多个标签的时候怎么办呢？应该选取哪个呢？谢谢！

**Comment by gaotianyu1350** [comment]
各个类别是分开评测的，所以有多个标签也没有问题。最后是对所有(h,r,t)的分数排序，然后计算AUC的

**Comment by ShellingFord221** [comment]
谢谢！test的时候我知道可以认为是多标签的，但是train的时候呢？因为要选取一个label作为这个bag的label

**Comment by gaotianyu1350** [comment]
train的时候如果一个pair有多个标签的话，会拆成多个bag，对于每个bag来说都是单标签的

**Comment by ShellingFord221** [comment]
你好，关于bag的representation vector的选取还想再请教一下。比如3个类别，这个bag中有2个句子，过softmax前的logit vector分别是
[a1, a2, a3]
[b1, b2, b3]
MAX的方法是选取每个类别最大的值构成一个新的vector，比如是[a1, b2, a3]
AVG的方法是选取每个类别的logit值求平均，为[(a1+b1)/2, (a2+b2)/2, (a3+b3)/2]
ATT的方法是每个类别的logit值*加权值，为[a1*w1+b1*w2, a2*w1+b2*w2, a3*w1+b3*w2]
最后将得到的新logit vector作为这个bag的representation vector，然后过softmax得到概率
但是这样有2个问题：
1. PCNN的代码在MAX部分好像是选取标签类最大的那个sample的logit vector作为这个bag的表示，比如这个bag的标签是3，而a3>b3，所以这个bag的表示向量就是[a1, a2, a3]
2. 直接用logit vector做比较的话有可比性吗？是不是要在比较前先用softmax normalize，然后得到representation vector(probability) 后再用softmax normalize?
非常感谢！

**Comment by gaotianyu1350** [comment]
1. 在训练阶段因为知道标签，所以可以这样做。测试阶段的话预测哪一类就取哪一类最大的。
2. 应该先用softmax，我们代码里面也是这么写的～

**Comment by ShellingFord221** [comment]
对于第2个问题，是否指的是：
instance_logit = tf.nn.softmax(__logit__(bag_hidden_mat, rel_tot), -1)
这一行将概率作为logit vector进行选取，然后模型最后还会再过一层softmax？

对于第1个问题，是否指的是：
训练时可以看到标签，所以选取[a1, a2, a3]，因为这个bag的标签是3且a3>b3（对应的代码是：
j = tf.argmax(instance_logit[:, query[i]], output_type=tf.int32)
bag_repre.append(bag_hidden_mat[j])）
测试时看不到标签，所以选取每个类别上最大的，即[a1, b2, a3]（对应的代码是：
bag_logit.append(tf.reduce_max(instance_logit, 0))）
但是这样会不会有train和test时选取标准不一致的问题呢？

非常感谢！！

**Comment by gaotianyu1350** [comment]
是这样的，至于train和test标准不一样，感觉问题不大吧


在2019年11月20日 20:30，ShellingFord221 写道：

对于第2个问题，是否指的是：
instance_logit = tf.nn.softmax(logit(bag_hidden_mat, rel_tot), -1)
这一行将概率作为logit vector进行选取，然后模型最后还会再过一层softmax？

对于第1个问题，是否指的是：
训练时可以看到标签，所以选取[a1, a2, a3]，因为这个bag的标签是3且a3>b3（对应的代码是：
j = tf.argmax(instance_logit[:, query[i]], output_type=tf.int32)
bag_repre.append(bag_hidden_mat[j])）
测试时看不到标签，所以选取每个类别上最大的，即[a1, b2, a3]（对应的代码是：
bag_logit.append(tf.reduce_max(instance_logit, 0))）
但是这样会不会有train和test时选取标准不一致的问题呢？

非常感谢！！

**Comment by ShellingFord221** [comment]
你好，关于train的时候bag的标签问题还想再请教下。在multi-instance one-label的情况下（即你们代码的实现），如果bag有多个标签，将其分别视为不同的sample。比如句子1,2,3,4,5属于一个bag，标签是a和b，那么就将其视为2个sample，即一个bag中句子是1,2,3,4,5，标签是a，另一个bag中句子也是1,2,3,4,5，但标签是b。可是这种方式不会使得模型confuse吗？sample长的一模一样，但是标签却不同，这种方式模型该如何根据类别提取特征呢？谢谢！

**Comment by ShellingFord221** [comment]
而且，训练的时候是每个bag指定一个标签，但测试的时候却是按照多分类的方法测试，这样是不是有点不太公平呢？测试的时候是不是只要预测对一个标签就可以了呢？（bag多个标签的情况下）
